api/envoy/type/matcher/v3/string.proto

```diff
@@ -61,6 +62,11 @@ message StringMatcher {
     //
     // * ``abc`` matches the value ``xyz.abc.def``
     string contains = 7 [(validate.rules).string = {min_len: 1}];
+
+    // Use an extension as the matcher type.
+    // [#extension-category: envoy.matching.string]
+    // [#comment: This uses the xds type instead of the envoy type to avoid a circular dependency in the build system.]
```

Comment:
We probably should in general be using this anyway, Should we have some CI format check to ensure we don't use stale xDS types  @envoyproxy/api-shepherds?

Reply:
Hmmm, there also a `xds.type.matcher.v3.StringMatcher` in xds repo, should we also update it?

@htuch Now the stale xDS types are used in almost everywhere. And there are lots of new APIs  (which are not synced to xds repo) in the `envoy/config/core` and `envoy/type/matcher`, so I think we should re-think does the xds repo still has actual meaning in practice.

cc @envoyproxy/api-shepherds

Reply:
I believe that our intention was to migrate to the types in the `xds` namespace over time.  I think we probably do need to enforce this somehow, or it will never happen.  We should stop allowing new instances of the old types in the `envoy` namespace.  We should also consider saying that new fields should be added only to the `xds` types, as a way of promoting migration: if a given call site needs a new field, that's an opportunity to add a field for the `xds` matcher type as an alternative to the existing `envoy` type.

Reply:
@wbpcode yes I think it makes sense to also add it to the xds StringMatcher type; that can be done pretty easily later.

Given how pervasive both the StringMatcher and TypedExtensionConfig are in the envoy namespace in the envoy config, I don't think we're ever going to be able to migrate existing usage to the xds namespace. It would break literally every envoy config, so I think we'll have to support using both. Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Reply:
I think it must be a reason that force us to use do this migration? But I didn't get it.

As a developer of Envoy, I think it's much easier to use/update the APIs under Envoy. Or we always to update another repo (I mean the cncf/xds repo).
As a reviewer, I think it would be more friendly to make the API and the implementation in same PR.
As a developer of control plane (a private fork of istio), I always use the `envoyproxy/go_control_plane` and never care the `cncf/xds`.
As a user of Envoy, I hope all APIs could be placed in the same position.
And I believe even as a developer of another xds client (like grpc), it's impossible to just use the `cncf/xds` only because it's simple and lack of good maintenance (compared to `envoy/api`).

In conclusion, I don't get a point why the migration is necessary. Anyone know the actual background and reason?

cc @envoyproxy/api-shepherds @envoyproxy/senior-maintainers

Reply:
> It would break literally every envoy config, so I think we'll have to support using both. Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Yeah. I agree with that migration is impossible.

And I also inclined to use `envoy.config.core.v3.TypedExtensionConfig` here (`envoy.type.matcher.v3.StringMathcer`) directly and use `xds.core.v3.TypedExtensionConfig` in `xds.type.matcher.v3.StringMatcher`. If we actually want to use the xds one to replace the envoy one, we should use the whole `xds.type.matcher.v3.StringMacher` to replace `envoy.type.matcher.v3.StringMatcher`.

It's weird for me to add a new field that use the xds API here because I think the `StringMatcher` is a very basic type.

The circular dependency is a problem, but maybe we could resolve in other way.

Reply:
>  Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Agree, we will followup at some point with tooling to enforce this as presubmit.

@wbpcode we will have a proposal coming up to move more things to the xDS repository. This would enable multi data plane consumption of xDS in a credible way with proper maintainership and governance. We will bring this to a GitHub issue and Envoy community meeting in the near future. I 100% appreciate the concerns you raise and I'm happy to do an offline 1:1 on this topic - I've started on a one pager to share for Envoy maintainers.

In the meanwhile, we should probably maintain our existing posture, which is to prefer these the cncf/xds base types where possible, so at least in this PR, we're all good. 

Thanks for the discussion and input.